File: problems/0315-count-of-smaller-numbers-after-self/solution.py
# Link: https://leetcode.cn/problems/count-of-smaller-numbers-after-self/
# Level: Hard
# Date: 2025-04-15

class FenwickTree:
    def __init__(self, size):
        self.n = size
        self.tree = [0] * (self.n + 1)

    def update(self, index, delta):
        while index <= self.n:
            self.tree[index] += delta
            index += index & -index
        
    def query(self, index):
        res = 0
        while index > 0:
            res += self.tree[index]
            index -= index & -index
        return res

class Solution:
    def countSmaller(self, nums: List[int]) -> List[int]:
        if not nums:
            return []
        sorted_num = sorted(nums)
        unique = []
        prev = None
        for num in sorted_num:
            if num != prev:
                unique.append(num)
                prev = num
        
        m = len(unique)
        ft = FenwickTree(m)
        counts = [0] * len(nums)
        for i in range(len(nums) - 1, -1, -1):
            x = nums[i]
            index = bisect.bisect_left(unique, x) + 1
            count = ft.query(index - 1)
            counts[i] = count
            ft.update(index, 1)
        return counts

File: problems/0315-count-of-smaller-numbers-after-self/solution.go
// Link: https://leetcode.cn/problems/count-of-smaller-numbers-after-self/
// Level: Hard
// Date: 2025-04-15

type FenwickTree struct {
	tree []int
	n    int
}

func NewFenwickTree(size int) *FenwickTree {
	return &FenwickTree{
		tree: make([]int, size+1),
		n:    size,
	}
}

func (ft *FenwickTree) Update(index int, delta int) {
	for index <= ft.n {
		ft.tree[index] += delta
		index += index & -index
	}
}

func (ft *FenwickTree) Query(index int) int {
	sum := 0
	for index > 0 {
		sum += ft.tree[index]
		index -= index & -index
	}
	return sum
}

func countSmaller(nums []int) []int {
	if len(nums) == 0 {
		return []int{}
	}

	sorted := make([]int, len(nums))
	copy(sorted, nums)
	sort.Ints(sorted)

	unique := []int{sorted[0]}
	prev := sorted[0]
	for i := 1; i < len(sorted); i++ {
		if sorted[i] != prev {
			unique = append(unique, sorted[i])
			prev = sorted[i]
		}
	}
	m := len(unique)
	ft := NewFenwickTree(m)
	res := make([]int, len(nums))
	for i := len(nums) - 1; i >= 0; i-- {
		x := nums[i]
		index := sort.Search(len(unique), func(j int) bool {
			return unique[j] >= x
		}) + 1
		count := ft.Query(index - 1)
		res[i] = count
		ft.Update(index, 1)
	}
	return res
}
